src/rt-app_parse_config.c

```diff
@@ -983,37 +983,94 @@ parse_task_phase_data(struct json_object *obj,
 {
 	/* used in the foreach macro */
 	struct lh_entry *entry; char *key; struct json_object *val; int idx;
-	int i;
+	struct json_object *array_val;
+	size_t i;
+	struct json_object *events_array = NULL;
+	bool is_first_entry;
 
 	log_info(PFX "Parsing phase");
 
+	data->nbevents = 0;
+	data->events = NULL;
+
 	/* loop */
 	data->loop = get_int_value_from(obj, "loop", TRUE, 1);
 
-	/* Count number of events */
-	data->nbevents = 0;
 	foreach(obj, entry, key, val, idx) {
-		if (obj_is_event(key))
+		if(!strncmp("events", key, 6)) {
+			if(!json_object_is_type(val, json_type_array)) {
+				log_critical(PIN "\"events\" key must be an array");
+				exit(EXIT_INV_CONFIG);
+			}
+			events_array = val;
+			break;
+		}
+	}
+
+	foreach(obj, entry, key, val, idx) {
+		if (obj_is_event(key)) {
+			/* Check that we are not trying to mix old and new styles */
+			if (events_array) {
```

Comment:
if an events array is detected we should simply ignore old style

Reply:
Is it really an improvement though ? The foreach is still needed to count the number of events anyway and silently ignoring events is not really something users would thank us for.

Reply:
but when we have found an array we can skip this counting entirely because we don't care about old style events

Reply:
My point is that this code cannot be removed, if someone is using old-style, this foreach needs to be somewhere. Silently ignoring part of the configuration file in order to save one iteration over a map that has a few tens of keys at most does not sound like a good deal. Even more considered the fact that in well-formed JSON files, the number of keys is bounded by the number of settings, so the only ones "paying a performance cost" are people writing a conf file that is ambiguous and will result in an error. (bearing in mind, we are parsing a configuration file, ahead of time of execution, in C, with an order of magnitude of a few tens or hundreds of keys in extreme cases).

On top of that, this can allow in the future to reclaim some event names for some settings. This sounds a bit crazy given that they have "verb" names for the most part, but the current syntax is lenient: we only match the prefix. This means that something like "synchronous-io" key is currently impossible to add, but would be possible with the "events" array as long as we forbid events outside of it when in use.